### PR TITLE
cache-to-.tar.zst

### PR DESCRIPTION
**Pull request by nathanhammond**
In 2022 gzip is almost always a bad choice. This PR starts from where #1991 finished, adopts Zstandard for compression, and supports restoration from both compressed and uncompressed caches.

The adopted zstd library requires CGO. This also edits our Makefile to reflect that requirement.

### CODE REVIEW

**Comment by nathanhammond** [comment]
The changes since the gzip version are here:

https://github.com/vercel/turborepo/pull/2195/files/810f369a3551f317e3abafcfec0b900c04fb76ab..3a6e987eb08ab5a661d0e506a606a59ef38f0cfa

File: cli/internal/cache/cache_http.go [hunk 1]
```diff
@@ -83,10 +84,11 @@ func (cache *httpCache) Put(anchor turbopath.AbsoluteSystemPath, hash string, du
...
 	defer w.Close()
-	gzw := gzip.NewWriter(w)
-	defer gzw.Close()
-	tw := tar.NewWriter(gzw)
-	defer tw.Close()
+	defer func() { _ = w.Close() }()
+	zw := zstd.NewWriter(w)
+	defer func() { _ = zw.Close() }()
+	tw := tar.NewWriter(zw)
+	defer func() { _ = tw.Close() }()
```
Comment (nathanhammond) [inline cli/internal/cache/cache_http.go:91]: This was lint cleanup. No material change, though we should probably fix this.

File: cli/internal/cache/cache_http.go [hunk 2]
```diff
@@ -256,12 +258,10 @@ func (cache *httpCache) retrieve(hash string) (bool, []turbopath.AnchoredSystemP
...
 	missingLinks := []*tar.Header{}
-	gzr, err := gzip.NewReader(reader)
-	if err != nil {
-		return nil, err
-	}
-	defer func() { _ = gzr.Close() }()
-	tr := tar.NewReader(gzr)
+	zr := zstd.NewReader(reader)
+	var closeError error
+	defer func() { closeError = zr.Close() }()
```
Comment (nathanhammond) [inline cli/internal/cache/cache_http.go:263]: Lint fixup, now we use this error.

File: cli/internal/cacheitem/create_test.go [hunk 3]
```diff
@@ -143,56 +143,63 @@ func TestCreate(t *testing.T) {
 		},
 	}
 	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			inputDir := turbopath.AbsoluteSystemPath(t.TempDir())
-			archiveDir := turbopath.AbsoluteSystemPath(t.TempDir())
-			archivePath := turbopath.AnchoredSystemPath("out.tar.gz").RestoreAnchor(archiveDir)
+		getTestFunc := func(compressed bool) func(t *testing.T) {
```
Comment (nathanhammond) [inline cli/internal/cacheitem/create_test.go:146]: Test the full set of behaviors from compressed and uncompressed files.

File: cli/internal/cacheitem/restore_test.go [hunk 4]
```diff
@@ -34,13 +35,12 @@ type restoreFile struct {
 func generateTar(t *testing.T, files []tarFile) turbopath.AbsoluteSystemPath {
 	t.Helper()
 	testDir := turbopath.AbsoluteSystemPath(t.TempDir())
-	testArchivePath := testDir.UntypedJoin("out.tar.gz")
+	testArchivePath := testDir.UntypedJoin("out.tar")
```
Comment (nathanhammond) [inline cli/internal/cacheitem/restore_test.go:38]: This now explicitly only creates a tar.

**Review by nathanhammond — COMMENTED** [review]
Some guides through the code changes.

**Review by gsoltis — APPROVED** [review]

**Comment by nathanhammond** [comment]
Merging over top of the linter complaint of (who knows what, it's giving a useless error message that I can't reproduce locally).

**Pull request merged**